Add support of `axis` for `tf.unique`

yongtang opened the pull request:
This fix tries to address the request from #11575 where `axis` was not supported for `tf.unique`. 

This fix adds support of `axis` for `tf.unique`. In this fix, the additional input `axis` has been specified as an `1D` vector so that it is possible to optionally provide an `axis` or not. In case `axis=[]` is provided, no axis is used. In case `axis=[x]` is used, `axis` for `tf.unique` is `x`. (ref  https://stackoverflow.com/questions/42754965/marking-input-as-optional-in-tensorflow)

This fix fixes #11575

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Convert unique implementation to be compatible with axis

This commit made the conversion so that unique could
be extended with axis. A custom unordered_map is used.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add support of `axis` for `tf.unique`

This commit adds support of `axis` for `tf.unique`.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test cases of `axis` support for `tf.unique`.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

drpngx commented:
@a-dai could you take a look?

caisq commented:
Gentle ping, @a-dai ?

a-dai commented on tensorflow/core/kernels/unique_op.cc:
```diff
@@ -33,48 +34,117 @@ class UniqueOp : public OpKernel {
-      auto it = uniq.insert(std::make_pair(Tin(i), j));
+    auto hash_fn = [&Tin](const int64& key) -> unsigned long {
+      size_t hash = 0;
+      for (int64 i = 0; i < Tin.dimension(0); i++) {
```
Some of the indices are int64 like here and some uint64 like in the line below, please make them consistent throughout.

yongtang replied:
Thanks. Done.

a-dai commented on tensorflow/core/kernels/unique_op.cc:
```diff
@@ -33,48 +34,117 @@ class UniqueOp : public OpKernel {
+      size_t hash = 0;
+      for (int64 i = 0; i < Tin.dimension(0); i++) {
+        for (uint64 j = 0; j < Tin.dimension(2); j++) {
+          hash += std::hash<T>{}(Tin(i, key, j));
```
This should be an ordered hash combiner or it could be expensive for big tensors.

yongtang replied:
Thanks. Done. The `Hash64Combine` from `tensorflow/core/lib/hash/hash.h` has been used.

a-dai reviewed (changes requested):
Nice extension, just a few comments.

yongtang committed:
Address review feedback.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented on tensorflow/core/ops/array_ops.cc:
```diff
@@ -2029,6 +2029,42 @@ y: 1-D.
 
+REGISTER_OP("UniqueV2")
+    .Input("x: T")
+    .Input("axis: int64")
```
Thanks @a-dai. The documentation has been updated.

yongtang commented:
Thanks @a-dai for the review. The PR has been updated. Please take a look and let me know if there are any issues.

a-dai commented:
Jenkins, test this please.

yongtang commented:
@caisq @a-dai @drpngx It seems that the Jenkins build is not starting automatically. Any chance to forceful start a Jenkins build so that the PR could be merged once Jenkins passes?

drpngx commented:
Jenkins, test this please.

drpngx commented:
Some kind of signature fail during docker build, it looks like. Jenkins, test this please.

yongtang commented:
The Jenkins tests passed. The other tests I think it will need to label the PR `kokoro:run`?

yongtang commented:
@gunan Not sure why some of the tests are failing on Jenkins as I could not reproduce on my dev machine. I based the PR and pushed again. Hopefully it will resolve some of the Jenkins issues. Can you relabel `kokoro:force-run` to give it a try again?

gunan commented:
Jenkins, test this please.

gunan commented:
It looks like the issues are related to this PR. Maybe a change in the behaviour of unique OP?

yongtang commented:
Thanks @gunan. Let me double check and see if I could fix it.

yongtang committed:
Change forward_input_or_allocate_output to allocate_output to fix

the following failures:
```
//tensorflow/python/debug:analyzer_cli_test
//tensorflow/python/debug:session_debug_file_test
//tensorflow/python/feature_column:feature_column_test
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@gunan I have updated the PR and fixed the issue. The issue is that `forward_input_or_allocate_output` should be changed to `allocate_output`.

I have tested locally and the following failed tests passes now:
```
//tensorflow/python/debug:analyzer_cli_test
//tensorflow/python/debug:session_debug_file_test
//tensorflow/python/feature_column:feature_column_test
```

Can you start the Jenkins again? Thanks a lot for the help.

gunan commented:
Jenkins, test this please.

shariq-audiofocus commented:
@a-dai is there a reason this hasn't been tested & merged yet?
